Update documentation

knutnergaard opened the pull request:
- add type annotations and edit/add documentation to `base/font.py`
- add new module for type annotation definitions: `base/type.py`
- add intersphinx mapping to python 3 inventory to `documentation/source/conf.py`

benkiel commented:
Looks like fontshell.font may need to be updated here: tests are run via fontShell, fyi. Though, not sure tbh if that is the issue.

knutnergaard commented:
I think the issue simply is that I changed the name of the new module from `types.py` to `type.py`, as per your request, right before pushing the commit to my local branch, but forgot to update the name in the import statement in `font.py`. Should I file a new PR or are you able to edit the file before merging?

knutnergaard commented:
The error was related to a misguided attempt of mine to fix a naming issue with the defaultLayer properties. I think the names of those are reversed, compared to their status as base vs. environment related. 

Anyway, reversing the changes I made, fixed the problem, and for fear of breaking  anything again, I haven't looked further into this, but you should definitely take a look.

benkiel commented:
Hey @knutnergaard, I am very sorry, I was typing far too quickly in my reply about the name for `types.py`: I do thing that `types.py` is better than `type.py`, the latter is going to be very confusing. Better to change it now before we merge.

knutnergaard commented:
> Hey @knutnergaard, I am very sorry, I was typing far too quickly in my reply about the name for `types.py`: I do thing that `types.py` is better than `type.py`, the latter is going to be very confusing. Better to change it now before we merge.

I had some trouble with the lasts few commits from the terminal, but hopefully everything is ok now. Sorry about that! 

I'll update `layer.py` once @typesupply gets back to me about my latest questions. :)

knutnergaard commented:
@benkiel, In `/base/layer.py` at line 53 (in the `_BaseGlyphVendor._setLayerInGlyph` method saying "Instance of '_BaseGlyphVendor' has no 'defaultLayer' member". This  missing attribute seemed like an oversight to me, so I just thought I'd mention it.

benkiel commented:
@knutnergaard I think that `_BaseGlyphVendor._setLayerInGlyph` bit is because `_BaseGlyphVendor` is subclassed by `BaseFont` which sets the logic for `defaultLayer`. Does that make sense?

knutnergaard commented:
> @knutnergaard I think that `_BaseGlyphVendor._setLayerInGlyph` bit is because `_BaseGlyphVendor` is subclassed by `BaseFont` which sets the logic for `defaultLayer`. Does that make sense?

Yes, that makes sense, but I see this error also occurs in other cases, like here, in the `BaseFont` class:

```python
def swapLayerNames(self, layerName: str, otherLayerName: str) -> None:
    ...
    self._swapLayers(layerName, otherLayerName) # Instance of 'BaseFont' has no '_swapLayers' member
```
I could not find any logic for `_swapLayers` in any other FontParts module.

benkiel commented:
`_swapLayers` is here: https://github.com/robotools/fontParts/blob/6446bb65573da81178c15e6a9dede1b68d7bf3a5/Lib/fontParts/base/font.py#L994

knutnergaard commented:
@benkiel @typesupply I've built a module to simplify the documentation process, according to the current style. I don't know where it fits in the current folder structure, but if you want to add it to the repository, I can make it available.

benkiel commented:
Hey @knutnergaard, I'll look at the documentation module — could you just share it as a gist right now?

Also, what do you think about making this PR against a new branch, `v1`. Better docs and typing would bring things to a spot where `1.0` feels right, and doing things as PRs against a branch would help others contribute to the typing effort, in addition to breaking up PRs into more easily reviewed chunks.

As for this PR — it will take me a bit to go through everything, I plan on having my review done by the 23rd. I know that's slow, but I have to do it between client projects and other work, so it's going to take a bit, FYI. I don't want you to think it's been forgotten (and I'll comment as I go)!

One thing that I see right now is the Sphinx build is throwing some errors, but it could be my version of Sphinx. I'll troubleshoot and come back with issues I can't solve.

Lastly — I would love to hear about how you're using fontParts in your work, the music font project seems really interesting!

A huge thank you for this work — it's really well done and I know it's a lot!

knutnergaard commented:
> Hey @knutnergaard, I'll look at the documentation module — could you just share it as a gist right now?
> 

Sure! It's [here](https://gist.github.com/knutnergaard/97b4b30df2f098a9495a300a46768276).

> Also, what do you think about making this PR against a new branch, `v1`. Better docs and typing would bring things to a spot where `1.0` feels right, and doing things as PRs against a branch would help others contribute to the typing effort, in addition to breaking up PRs into more easily reviewed chunks.
> 

Sounds good.

> As for this PR — it will take me a bit to go through everything, I plan on having my review done by the 23rd. I know that's slow, but I have to do it between client projects and other work, so it's going to take a bit, FYI. I don't want you to think it's been forgotten (and I'll comment as I go)!
> 

Thanks for the update. I realise that you have a day job, so no worries. :)

> One thing that I see right now is the Sphinx build is throwing some errors, but it could be my version of Sphinx. I'll troubleshoot and come back with issues I can't solve.
> 

I know. Please do.

> Lastly — I would love to hear about how you're using fontParts in your work, the music font project seems really interesting!
> 

I'm not doing much design work myself nowadays, but the library I've built, [SMufoLib](https://smufolib.readthedocs.io/en/latest/), extends the functionality of FontParts to aid in building fonts following the [SMuFL standard](https://w3c.github.io/smufl/latest/), which provides mapping and metadata standards for music fonts. Music fonts are very different from text fonts, in that the glyphs are much less uniform in nature. The work is also generally less involved with regard to spacing and kerning and more involved when it comes to balancing the size and weight of different symbols. Most importantly, SMufoLib utilises the ufo font lib and glyph lib objects to store and provide access to SMuFL-specific metadata. It does not implement FontParts in the intended way, but inherits directly from the `fontshell` objects. As such, it's completely console-based. 

Actually, using FontParts and the UFO format for this was kind of a last resort, due to the short-comings of the current FontLab scripting API(s). As it turns out, the lib objects provided excellent means for data storage which I originally did not consider when thinking about this library strictly in FontLab terms.

> A huge thank you for this work — it's really well done and I know it's a lot!

You're very welcome and thanks! :)

typesupply commented:
> I'm not doing much design work myself nowadays, but the library I've built, [SMufoLib](https://smufolib.readthedocs.io/en/latest/), extends the functionality of FontParts to aid in building fonts following the [SMuFL standard](https://w3c.github.io/smufl/latest/), which provides mapping and metadata standards for music fonts.

**Wow wow wow wow wow!** This is so cool! 🤩

benkiel commented:
@knutnergaard thanks for the flexibility! I changed this PR to be against a `v1` branch, so we can work there. I'll get on reviewing this PR and we can make separate PRs to this branch to add/do other things.

`SMufoLIb` is so cool! So glad that `lib` is doing exactly what we hoped they could be used for building interesting things for folks that wouldn't make sense in a general tool/spec!

benkiel commented:
(and thank you for understanding the speed of review!)

roberto-arista commented:
Hey there! I'd be happy to contribute and help with this process. I use `fontParts` exclusively in a IDE environment and I'd love to have annotations to improve my developer experience. What could be the best way to proceed? A separate PR for each `fontParts.base` module missing?

benkiel commented:
@roberto-arista Yes, PR for seperate things to the `v1` branch

benkiel commented:
@knutnergaard Ok — quick look at the documentation module — WOW!

I _think_ the best place to put it in the `documentation` folder, in a new folder, either `helpers` or `tools`. Doesn't make sense to put it in the main `fontParts` folder structure

knutnergaard commented:
> @knutnergaard Ok — quick look at the documentation module — WOW!
> 
> I _think_ the best place to put it in the `documentation` folder, in a new folder, either `helpers` or `tools`. Doesn't make sense to put it in the main `fontParts` folder structure

Great! I've put it in documentation > tools. I've done some minor refactoring, and improvements and added functionality to document normalization.

Also, I've added abstract versions of members used in the mixin but defined in the main `Doctring` class to avoid certain `mypy` errors. Now that annotation is introduced,  this might be relevant for `fontParts` as well.

knutnergaard commented:
@benkiel , after another pass at annotations in `base.py` and `glyphs.py`, everything passes when running `mypy` except for these errors:
```zsh
glyph.py:76: error: Signature of "_reprContents" incompatible with supertype "BaseObject"  [override]
glyph.py:76: note:      Superclass:
glyph.py:76: note:          @classmethod
glyph.py:76: note:          def _reprContents(cls) -> List[str]
glyph.py:76: note:      Subclass:
glyph.py:76: note:          def _reprContents(self) -> List[str]
glyph.py:2707: error: Signature of "isCompatible" incompatible with supertype "InterpolationMixin"  [override]
glyph.py:2707: note:      Superclass:
glyph.py:2707: note:          def isCompatible(self, other: Any, cls: Type[Any]) -> Tuple[bool, Any]
glyph.py:2707: note:      Subclass:
glyph.py:2707: note:          def isCompatible(self, other: BaseGlyph) -> Tuple[bool, str]
```
Is there any reason why these methods are defined on the class level in `BaseObject`?
I'll commit these changes once your initial review is done.

knutnergaard commented:
@benkiel, regarding `bPoint.py` can you supply the type descriptions in the docstring table of `BaseBPoint.type`? I'm not sure how to describe them, and I feel like they should have a more thorough description than `'corner'` or `'curve'`.

benkiel commented:
@knutnergaard Yes, will do. I had planned on reviewing this PR this morning/today, but...life. Know you are waiting on me, will get to it in short order.

benkiel commented:
@knutnergaard for `bPoint.py`:

corner - a point where bcpIn and bcpOut are not smooth (linked)
curve - a point with bcpIn and bcpOut are smooth (linked)

roberto-arista commented:
@knutnergaard I've seen that you started to work on `bPoint.py` in this PR. Should I close the other ones? Do you prefer work alone on this? Or, if it's helpful I can fork your repo and work on some other objects. Let me know.

knutnergaard commented:
> @knutnergaard I've seen that you started to work on `bPoint.py` in this PR. Should I close the other ones? Do you prefer work alone on this? Or, if it's helpful I can fork your repo and work on some other objects. Let me know.

Hey @roberto-arista I realise I'm not really well versed with GitHub branches and different pull requests. For some reason, all the edits done in the `master` branch of my own fork is just linked to this pull request, so I've just continued to push all my edits to `robotools:master` like before. 

Since I see you've created a dedicated branch for each module in your own fork, and  @benkiel has suggested we work within a `v1` branch, I thought I'd at least create a similar 'v1' branch in my fork, and then bring my own `master` up to date with `robotools:master`.  However, considering that this pull request now is set up to merge only 2 commits, rather than the original 37, that seems to have been not such a good idea, and I'm not really sure how to resolve the conflict. I think I have to read up on GitHub best practices. 🙈

I appreciate any help you can give with the annotation. My work on `bPoint.py` is based on your edits. I've just added documentation, basically. I'm more than willing to commit my edits to separate PRs, or even add them to the ones you've created. I'm just not sure how to. Sorry about that.

knutnergaard commented:
> @knutnergaard I've seen that you started to work on `bPoint.py` in this PR. Should I close the other ones? Do you prefer work alone on this? Or, if it's helpful I can fork your repo and work on some other objects. Let me know.

At present, all the commits to my own branch seem to update this PR automatically. I'm happy to use separate PRs like you've done, I'm just not sure how.

Anyway, my edits to bPoint.py is based on yours. I've basically just updated the documentation. I appreciate all the help you can give with the annotation, as I use those as basis when updating docstrings.

roberto-arista commented:
Ok, let's wait for @benkiel's input on this PR. Since you are doing also work on the docs, I agree that it might make more sense to contribute to your fork first.

knutnergaard commented:
> Ok, let's wait for @benkiel's input on this PR. Since you are doing also work on the docs, I agree that it might make more sense to contribute to your fork first.

Great! In that case, I'll create separate branches for annotations and docs, so we can work more cleanly than I've done so far.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -540,264 +693,390 @@ def kerningSortKeyFunc(pair):
 
             >>> font.layerOrder = ["My Layer 2", "My Layer 1"]
             >>> font.layerOrder
             ["My Layer 2", "My Layer 1"]
```
Should be `("My Layer 2", "My Layer 1")`

knutnergaard replied:
Fixed.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -806,33 +1085,46 @@ def getLayer(self, name):
         for layer in self.layers:
             if layer.name == name:
                 return layer
+        else:
```
This `else` feels wrong, there is no corresponding if. Should just be the `raise` from the next line

knutnergaard replied:
Indeed. Fixed, and added corresponding `:raises:` statement to docstring.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1115,78 +1517,106 @@ def __setitem__(self, name, glyph):
 
             >>> font.glyphOrder
             ["C", "B", "A"]
             >>> font.glyphOrder = ["A", "B", "C"]
```
actually returns a tuple, so should be `("A", "B", "C")`

knutnergaard replied:
Fixed.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1115,78 +1517,106 @@ def __setitem__(self, name, glyph):
     )
 
-    def _get_base_glyphOrder(self):
+    def _get_base_glyphOrder(self) -> List[str]:
```
Does type need to be Tuple here or is List good?

knutnergaard replied:
No, should be `tuple`. All of these methods were wrongly typed and documented, but are now fixed.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1455,10 +2004,7 @@ def _interpolate(self, factor, minFont, maxFont,
             dstLayer.interpolate(factor, minLayer, maxLayer,
                                  round=round, suppressError=suppressError)
         if self.layerOrder:
-            if ufoLib.DEFAULT_LAYER_NAME in self.layerOrder:
```
Very hesitant to change this, but see why it may be ok. @typemytype or @typesupply thoughts?

knutnergaard replied:
Is there a glitch somewhere? Can't find this line in my original `font.py` file pulled from `master`.

benkiel replied:
I see it: https://github.com/robotools/fontParts/blob/2bfa558107e9ab7c3e2ccfdec098a62dafae5784/Lib/fontParts/base/font.py#L1458

knutnergaard replied:
For some reason, `git pull` didn't work, and I had to reclone the repo to see these lines remotely, which is why they were missing. It was never my intention to delete them, and so I've added them to reflect [2bfa558](https://github.com/robotools/fontParts/commit/2bfa558107e9ab7c3e2ccfdec098a62dafae5784).

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1537,42 +2100,72 @@ def _isCompatible(self, other, reporter):
+        Example::
+
+            {
+            'A': {'Aacute', 'Aring'},
```
should be `()` not `{}`

knutnergaard replied:
Ghanged, I've also made the example more complete:

```python
"""
...

Example::

    >>> mapping = font.getReverseComponentMapping()
    >>> mapping
    {'A': ('Aacute', 'Aring'), 'acute': ('Aacute',),
    'ring': ('Aring',)  , ...}

"""
```

benkiel replied:
THANK YOU!

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1537,42 +2100,72 @@ def _isCompatible(self, other, reporter):
+        Note that one glyph can have multiple unicode values, and a
+        unicode value can have multiple glyphs pointing to it.
+
+        :return: A :class:`dict` of component glyph names mapped to sets
```
sets feels wrong here. List or tuple?

knutnergaard replied:
Changed.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1537,42 +2100,72 @@ def _isCompatible(self, other, reporter):
-    def getCharacterMapping(self):
-        """
-        Create a dictionary of unicode -> [glyphname, ...] mappings.
-        All glyphs are loaded. Note that one glyph can have multiple unicode values,
```
Lets not lose the note about multiple unicodes

knutnergaard replied:
Fixed.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -1583,116 +2176,180 @@ def _getCharacterMapping(self):
+            instances.
 
-        Getting selected guideline objects:
+        Getting selected guideline objects::
```
I do like this breaking up. Does it need an overall `Example` heading? If so, this would apply in other places here

benkiel reviewed (changes requested):
@knutnergaard finally forced myself to wade through it all (do not tell my business partner that this is why I am behind on some things...)

benkiel commented on Lib/fontParts/base/layer.py:
```diff
@@ -363,14 +507,24 @@ def _set_selectedGlyphNames(self, value):
-           "'%s'" % self.name,
+    def _reprContents(self) -> List[str]:
+        contents: List[str] = [
+            "'%s'" % self.name,
```
I have made a issue for moving everything to f-strings, so you can resolve and ignore this one

benkiel commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+        This is the environment implementation of :meth:`BaseDict.keys`.
+
+        :return: A :class:`list` of dictionary keys. If
+            a :cvar:`BaseDict.keyNormalizer` is set, it will have been applied
```
better to say `it will be applied to each key in the calling method`

knutnergaard replied:
Changed.

benkiel commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+
+        :return: A :class:`list` of :class:`tuple` items containing key-value pairs.
+            If both :cvar:`BaseDict.keyNormalizer` and :cvar:`BaseDict.valueNormalizer`
+            are set, they will have been applied to the keys and values, respectively.
```
better to say `they will be applied in the calling method to the key and values`

knutnergaard replied:
Changed.

benkiel commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+            a :cvar:`BaseDict.keyNormalizer` is set, it will have been applied
+            to the key.
+        :return: The value for the given key. If
+            a :cvar:`BaseDict.valueNormalizer` is set, it will have been applied
```
better to say `it will be applied in the calling method to the returned value`

knutnergaard replied:
Changed.

benkiel commented on Lib/fontParts/base/base.py:
```diff
@@ -364,26 +577,66 @@ def get(self, key, default=None):
+        :param default: The default value to return if the key is not found.
+        :return: The value associated with the given key, or the default value
+            if the key is not found. If a :cvar:`BaseDict.valueNormalizer` is set,
+            it will have been applied to the returned value.
```
better to say `it will be applied in the calling method to the returned value`

knutnergaard replied:
Changed.

benkiel commented on Lib/fontParts/base/base.py:
```diff
@@ -393,75 +646,132 @@ def pop(self, key, default=None):
+        :param default: The default value to return if the key is not found.
+        :return: The value associated with the given key, or the default value
+            if the key is not found. If a :cvar:`BaseDict.valueNormalizer` is set,
+            it will have been applied to the returned value.
```
better to say `it will be applied in the calling method to the returned value`

knutnergaard replied:
Changed.

benkiel reviewed (changes requested):
Changed mind about nomalizer text in docs, keeping to private methods by changing tense

knutnergaard commented on Lib/fontParts/base/font.py:
```diff
@@ -210,85 +275,103 @@ def save(self, path=None, showProgress=False, formatVersion=None, fileStructure=
-        +--------------+--------------------------------------------------------------------+
-        | unixascii    | UNIX ASCII font (ASCII/PFA)                                        |
-        +--------------+--------------------------------------------------------------------+
+    def generateFormatToExtension(format: str, fallbackFormat: str) -> str:
```
Sorry, not sure how I missed that. Will fix, but isn't this format and placement slightly better:

```python
"""
:param format: The format identifier to map to a file extension. Options are:

        +--------------------+-------------------------------------------------------+
        | Format             | Description                                           |
        +--------------------+-------------------------------------------------------+
        | ``'mactype1'``     | Mac Type 1 font (generates suitcase  and LWFN file)   |
        | ``'macttf'``       | Mac TrueType font (generates suitcase)                |
        | ``'macttdfont'``   | Mac TrueType font (generates suitcase with data fork) |
        | ``'otfcff'``       | PS OpenType (CFF-based) font (OTF)                    |
        | ``'otfttf'``       | PC TrueType/TT OpenType font (TTF)                    |
        | ``'pctype1'``      | PC Type 1 font (binary/PFB)                           |
        | ``'pcmm'``         | PC MultipleMaster font (PFB)                          |
        | ``'pctype1ascii'`` | PC Type 1 font (ASCII/PFA)                            |
        | ``'pcmmascii'``    | PC MultipleMaster font (ASCII/PFA)                    |
        | ``'ufo1'``         | UFO format version 1                                  |
        | ``'ufo2'``         | UFO format version 2                                  |
        | ``'ufo3'``         | UFO format version 3                                  |
        | ``'unixascii'``    | UNIX ASCII font (ASCII/PFA)                           |
        +--------------------+--------------------------------------–––------------–-+
"""

knutnergaard commented on Lib/fontParts/base/layer.py:
Not intentionally. I'll keep them all.

knutnergaard commented on documentation/tools/docstring.py:
```diff
@@ -0,0 +1,964 @@
+    'important': '.. important::',
+    'seealso': '.. seealso::',
+    'tip': '.. tip::',
+    'todo': '..todo::'
```
Indeed! Fixed.

knutnergaard commented on documentation/tools/docstring.py:
Do you find the module docstring unclear?
Agree that examples should be provided for each class and function, but maybe not at the module level?

knutnergaard commented on documentation/source/objectref/objects/layer.rst:
```diff
@@ -150,3 +183,20 @@ Environment
 
 .. automethod:: BaseLayer.naked
 .. automethod:: BaseLayer.changed
+
```
No, deleted.

knutnergaard commented on Lib/fontParts/base/bPoint.py:
```diff
@@ -273,20 +460,49 @@ def _set_bcpOut(self, value):
+        | type   | Description                                             |
+        +--------+---------------------------------------------------------+
+        | curve  | A point where bcpIn and bcpOut are smooth (linked).     |
+        | corner | A point where bcpIn and bcpOut are not smooth (linked). |
```
Fixed. I suggest clarifying that the type is a string, in line with comments above, like this:

```python
"""
+--------------+-----------------------------------------------------------+
| Type         | Description                                               |
+--------------+-----------------------------------------------------------+
| ``'curve'``  | A point where bcpIn and bcpOut are smooth (linked).       |
| ``'corner'`` | A point where bcpIn and bcpOut are not smooth (unlinked). |
+--------+-----------------------------------------------------------------+
"""
```
Do you agree?

benkiel replied:
Yes, that looks great

knutnergaard commented on Lib/fontParts/base/point.py:
```diff
@@ -295,29 +450,40 @@ def _set_y(self, value):
+        """Get the index of the point.
+
+        :return: An :class:`int` representing the point's index within an
+            ordered list of the parent glyph's points.
```
Isn't this clear from the summary line stating that it is only a getter?

Anyway, I'll change it.

benkiel replied:
Yes, but I'm thinking of folks who might not understand from only having a `get` that you can't set the value. Good to be explicit.

knutnergaard replied:
I see. Changed.

Should all exclusive getters have a read-only note?

benkiel replied:
Yes, they should

knutnergaard replied:
Added.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -9,191 +14,255 @@
-        **name** will be a :ref-type-string` representing
-        a glyph name. It will have been normalized with
+        :meth:`BaseLayer.__contains__` and :meth:`BaseFont.__contains__`.
         :func:`normalizers.normalizeGlyphName`.
```
Can't find the line in my current version of the file so, I must have spotted it post commit.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -206,71 +275,112 @@ def newGlyph(self, name, clear=True):
+
+        :param name: The name of the glyph to remove.
 
-        This method is deprecated. :meth:`BaseFont.__delitem__` instead.
```
I've moved this into a dedicated deprecated directive at the top, according to prior discussion (as far as I can remember):

```python

 """Insert a specified glyph into the layer.

        .. deprecated::

            This method is deprecated. Use :meth:`BaseFont.__setitem__` instead.
...

"""
```
Should I keep the old format?

benkiel replied:
No, don't keep the old format, thank you!

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -762,41 +1079,80 @@ def _isCompatible(self, other, reporter):
+        tuples of glyph names. Each Unicode value corresponds to one or
+        more glyphs, and the glyph names represent these glyphs in the
+        mapping.
+
```
Should this maybe be placed in a note directive or is it not important enough?

benkiel replied:
Good to have it as a note directive, good call

knutnergaard replied:
Done!

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -139,109 +193,167 @@ def _get_font(self):
+        :attr:`BaseGlyph.unicodes` to a :class:`tuple` containing that value,
+        or to an empty :class:`tuple` if value is :obj:`None`.
+
+        The value must be an :class:`int` or :obj:`None`.
```
Again, a miss, but see my comment above with regard to the correct referencing.

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -660,33 +945,67 @@ def drawPoints(self, pen, contours=True, components=True):
-        - guidelines
-        - image
-
-        It's possible to turn off the clearing of portions of
```
Should it go above the parameter section (my preference) or as an introduction to the second line in the example?

benkiel replied:
Ok to have in the parameter section

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -1483,53 +2134,81 @@ def _round(self):
           (such threshold is calculated as the smallest contour width
           or height in the glyph divided by two)
-        - the (negative) surface of the bounding box of the contour: width * height
-        the latter is a safety net for for instances like a very thin 'O' where the
```
A copy paste error. I've moved the missing note to a following section and capitalised the first word.

What does `_negative_` in this missing note refer to? should it be written as inline code ``_negative_`` or are the underlines there for emphasis?

benkiel replied:
there for emphasis

knutnergaard replied:
In that case, perhaps it's better to use bold or italic text? Underlined text isn't supported in reST, at least not without css tweaking.

benkiel replied:
Agree, just need emphasis of either bold or italic. My preference here is italic.

knutnergaard replied:
Agree. Fixed.

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -2051,18 +2961,28 @@ def _get_area(self):
-        Immutable tuple of the glyph's layers.
+        """Get the layers of the glyph.
+
+        :return: A :class:`tuple` of the :class:`BaseLayer` instances belonging
```
Isn't this clear from the return section?

If not, would ```Get the layers of the glyph as a :class:`tuple`.``` suffice? After all, "immutable tuple" is redundant.

benkiel replied:
sorry, agree, resolving

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -2152,70 +3100,104 @@ def removeLayer(self, layer):
+        :raises IOError: If no valid image file can be found at the given path.
+        :raises FontPartsError:  If `path` and `data` are both provided.
+
+        Add the image as a filepath::
```
No, I think we should only use `Example::` where there is no introductory text needed. Otherwise we should style them all as a heading with a custom directive (i.e.,  `:example:`) or strong formatting `**Example::**`.

benkiel replied:
Agree, thank you!

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -95,108 +89,177 @@ def __get__(self, obj, cls):
         """
         equal = self.__eq__(other)
-        return NotImplemented if equal is NotImplemented else not equal
+        return False if equal is NotImplemented else not equal
```
Sorry, this was a solution to `mypy` throwing an error since the return type is only expected to be a `bool`. It was not correct, though. Simply refactoring the return logic into a multi-liner solved it.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -9,191 +14,255 @@
+        Example::
 
-            >>> glyph = layer["A"] = otherGlyph
+            >>> glyph = font.insertGlyph(otherGlyph, name="glyph2")
```
fixed!

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -762,41 +1079,80 @@ def _isCompatible(self, other, reporter):
+        the composite glyph names that include the comoponent. All
+        glyphs are loaded.
+
+        Note that one glyph can have multiple unicode values, and a
```
Fixed!

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -139,109 +193,167 @@ def _get_font(self):
+        property getter.
+
+        :return A :class:`str` defining the name of the glyph. The value
+            will benormalized with :func:`normalizers.normalizeLayerName`.
```
Fixed!

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -872,407 +1262,586 @@ def _appendContour(self, contour, offset=None, **kwargs):
-        This must return an integer indicating
-        the number of components in the glyph.
+    def _lenComponents(self, **kwargs: Any) -> int:
+        r"""Get the number of components in the ntive glyph.
```
Fixed!

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -1,57 +1,51 @@
-            foo = property(_get_foo, _set_foo)
+    .. code-block:: python
 
-
```
Reinstated and assigned to code blocks.

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -1281,139 +1850,195 @@ def _clearAnchors(self):
             >>> guidelines = glyph.guidelines
 
-        The list will contain :class:`BaseGuideline` objects.
+        The list will contain  objects.
```
Actually, I don't think the line should be there at all, given the current format with a dedicated `:return:` section. I must have neglected deleting the line after copying the role.

benkiel replied:
Got it, feel free to delete line, happy with that.

knutnergaard replied:
Done!

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -2467,33 +3573,52 @@ def _dumpToGLIF(self, glyphFormatVersion):
+        Example::
 
-            >>> contours = glyph.selectedContours:
+            >>> components = glyph.selectedContours
```
Fixed.

benkiel commented on Lib/fontParts/base/font.py:
```diff
@@ -105,96 +142,124 @@ def copyData(self, source):
+        of :meth:`BaseFont.__init__`. It should wrap a native font
+        object based on the value type of the `pathOrObject` parameter.
 
-        +--------------------+---------------------------------------------------+
```
Great, thank you

benkiel commented on Lib/fontParts/base/glyph.py:
```diff
@@ -139,109 +193,167 @@ def _get_font(self):
-        The glyph's unicode values in order from most to least important.
+        """Get or set the glyph's Unicode values.
+
+        The value must be a :class:`list` or :class:`tuple` of :class:`int`
```
Yes, that's good.

knutnergaard commented on Lib/fontParts/base/glyph.py:
```diff
@@ -1712,17 +2504,26 @@ def _fromMathGlyph(self, mathGlyph, toThisGlyph=False, filterRedundantPoints=Tru
             if identifier is not None:
                 g._setIdentifier(identifier)
         copied.lib.update(mathGlyph.lib)
-        if not toThisGlyph:
```
Unintentional. Fixed.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -95,108 +89,177 @@ def __get__(self, obj, cls):
+    :param v: The factor value determining the interpolation between `a` and `b`
+        as a single :class:`int` or :class:`float` or a :class:`tuple` of
+        two :class:`int` or :class:`float` values representing the factors
+        ``(x, y)``.
```
Done! Would it be ok to rename these parameters to something more descriptive?

benkiel replied:
Yes: I'm ok with renaming the parameters to be more clear

knutnergaard replied:
Done!

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+    def changed(self, *args: Any, **kwargs: Any) -> None:
+        r"""Tell the environment that something has changed in the object.
 
-            >>> obj.changed()
```
Agree. Reinstated.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
-        Return the environment's native object
-        that has been wrapped by this object.
 
-            >>> loweLevelObj = obj.naked()
```
Agree. Reinstated.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
         """
         self.raiseNotImplementedError()
 
-    has_key = __contains__
```
Unintentional. Reinstated.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -393,75 +646,132 @@ def pop(self, key, default=None):
+        :param other: An object of key-value pairs to update this dictionary with.
+
+        """
+        from copy import deepcopy
```
Might I be working with an old file perhaps (see my comment about having to reclone the repo)?
Anyway, not my intention to move it there, so it's back at the top.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -393,75 +646,132 @@ def pop(self, key, default=None):
                 value = self.valueNormalizer.__func__(value)
                 d[key] = value
-            value = d
+            other = d
```
Indeed, or rather:
```python
otherCopy = deepcopy(other)
if self.keyNormalizer is not None and self.valueNormalizer is not None:
    d = {}
    for key, value in otherCopy.items():
        key = self.keyNormalizer.__func__(key)
        value = self.valueNormalizer.__func__(value)
        d[key] = value
    value = d
self._update(otherCopy)
```
Anyway, fixed.

knutnergaard commented:
> Changed mind about nomalizer text in docs, keeping to private methods by changing tense

Just to be clear, should these references consistently be in the future tense in all modules?

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -659,107 +1024,171 @@ def isCompatible(self, other, cls):
+
+        .. important::
+
+            Subclasses must override this method.
```
Fixed.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -768,100 +1197,139 @@ def _set_position(self, value):
+    def _setIdentifier(self, value: str) -> None:
+        """Force a specific identifier onto an object.
+
+        THis method is intended for subclasses that allow setting an
```
Fixed.

benkiel commented:
@knutnergaard yes, future tense when the value obtained by the private method will be normalized in the public method, past tense when the value given to the private method has been normalized by the public method before being handed to the private method.

Hopefully that is clear (I know, confusing)

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -659,107 +1024,171 @@ def isCompatible(self, other, cls):
+        the :attr:`PointPositionMixin.position` property getter.
+
+        :return: The current point position as a :ref:`type-coordinate`.
+            The value will have been normalized
```
Changed.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -476,160 +786,201 @@ def transformBy(self, matrix, origin=None):
+
+        This is the environment implementation of :meth:`BaseObject.scaleBy`.
+
+        :param value: The x and y values to scale the glyph by as
```
Fixed.

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+        This is the environment implementation of :meth:`BaseDict.values`.
+
+        :return: A :class:`list` of dictionary values. If
+            a :cvar:`BaseDict.valueNormalizer` is set, it will have been applied
```
Changed.

knutnergaard commented:
Apart from adding documentation https://github.com/robotools/fontParts/pull/739/commits/9e869d55150b5ed9c55a450837dc888e1106a365 also adds type annotation where missing and improves code generally.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -509,6 +568,15 @@ def _set_selectedGlyphNames(self, value: CollectionType[str]) -> None:
 
     has_key: Callable[[_BaseGlyphVendor, str], bool] = __contains__
 
+    # ----------------
```
Removed.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -1,6 +1,7 @@
 # pylint: disable=C0103, C0302, C0114, W0613
 from __future__ import annotations
 from typing import TYPE_CHECKING, Any, Callable, Iterator, List, Optional, Tuple
+from abc import ABC, abstractmethod
```
Removed.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -509,6 +568,15 @@ def _set_selectedGlyphNames(self, value: CollectionType[str]) -> None:
 
     has_key: Callable[[_BaseGlyphVendor, str], bool] = __contains__
 
+    # ----------------
+    # Abstract Members
+    # ----------------
+
+    @property
+    @abstractmethod
+    def defaultLayer(self) -> BaseLayer:
+        """Subclasses must implement this attribute."""
+
```
Removed.

benkiel commented on Lib/fontParts/base/glyph.py:
```diff
@@ -1493,7 +1491,7 @@ def appendComponent(self,
                     identifier = normalizedComponent.identifier
-        baseGlyph = normalizers.normalizeGlyphName(baseGlyph)
+        normalizedBaseGlyph = normalizers.normalizeGlyphName(baseGlyph)
         if self.name == baseGlyph:
```
should be `normalizedBaseGlyph`

knutnergaard replied:
Fixed.

benkiel reviewed (changes requested):
Just a couple of things!

knutnergaard commented on Lib/fontParts/base/base.py:
```diff
@@ -704,16 +754,15 @@ def update(self, other: BaseDict) -> None:
                 d[key] = value
-            other = d
-        self._update(other)
+            value = d
```
I agree that `value = d` doesn't seem to make much sense, but I'm also a bit confused as to what should be passed to `_update`.

To make it clear, the current logic is:
```python
otherCopy = deepcopy(other)
if self.keyNormalizer is not None and self.valueNormalizer is not None:
    d = {}
    for key, value in otherCopy.items():
        key = self.keyNormalizer.__func__(key)
        value = self.valueNormalizer.__func__(value)
        d[key] = value
    value = d
self._update(otherCopy)
```
This should be changed to:
```python
otherCopy = deepcopy(other)
if self.keyNormalizer is not None and self.valueNormalizer is not None:
    d = {}
    for key, value in otherCopy.items():
        key = self.keyNormalizer.__func__(key)
        value = self.valueNormalizer.__func__(value)
        d[key] = value
    other = d # other here
self._update(other) # and here
```
Correct?

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -116,14 +133,14 @@ def __getitem__(self, name: str) -> BaseGlyph:
         return glyph
 
-    def _getItem(self, name: str, **kwargs: Any) -> BaseGlyph:
+    def _getItem(self, name: str, **kwargs: Any) -> BaseGlyph:  # type: ignore[return]
```
Yes, this is for `mypy` to ignore the return value since the method is a placeholder and doesn't actually return anything.

knutnergaard commented on Lib/fontParts/base/layer.py:
```diff
@@ -180,10 +203,15 @@ def keys(self) -> Tuple[str, ...]:
         return self._keys()
 
-    def _keys(self, **kwargs: Any) -> Tuple[str, ...]:
+    def _keys(self, **kwargs: Any) -> Tuple[str, ...]:  # type: ignore[return]
```
Indeed. Se comment above. 

This isn't a very elegant solution, but the least intrusive one I could come up with after some research.

benkiel commented:
APPROVED! Thank you so much for all this work, I know it is _a lot_!
